plugins/Cors/CorsPlugin.py

```diff
@@ -74,28 +75,56 @@ def actionOptionalFileInfo(self, to, inner_path, *args, **kwargs):
         return self.corsFuncWrapper("actionOptionalFileInfo", to, inner_path, *args, **kwargs)
 
     def actionCorsPermission(self, to, address):
-        site = self.server.sites.get(address)
-        if site:
-            site_name = site.content_manager.contents.get("content.json", {}).get("title", address)
-            button_title = _["Grant"]
+        if isinstance(address, list):
+            addresses = address
         else:
-            site_name = address
-            button_title = _["Grant & Add"]
-
-        if site and "Cors:" + address in self.permissions:
+            addresses = [address]
+
+        button_title = _["Grant"]
+        site_names = []
+        site_addresses = []
+        for address in addresses:
+            site = self.server.sites.get(address)
+            if site:
+                site_name = site.content_manager.contents.get("content.json", {}).get("title", address)
+            else:
+                site_name = address
+                # If at least one site is not downloaded yet, show "Grant & Add" instead
+                button_title = _["Grant & Add"]
+
+            if not (site and "Cors:" + address in self.permissions):
+                # No site or no permission
+                site_names.append(cgi.escape(site_name))
+                site_addresses.append(address)
+
+        if len(site_names) == 0:
             return "ignored"
 
         self.cmd(
             "confirm",
-            [_["This site requests <b>read</b> permission to: <b>%s</b>"] % html.escape(site_name), button_title],
-            lambda res: self.cbCorsPermission(to, address)
+            [_["This site requests <b>read</b> permission to: <b>%s</b>"] % ", ".join(map(html.escape, site_names)), button_title],
+            lambda res: self.cbCorsPermission(to, site_addresses)
         )
 
-    def cbCorsPermission(self, to, address):
-        self.actionPermissionAdd(to, "Cors:" + address)
-        site = self.server.sites.get(address)
-        if not site:
-            self.server.site_manager.need(address)
+    def cbCorsPermission(self, to, addresses):
+        # Add permissions
+        for address in addresses:
+            permission = "Cors:" + address
+            if permission not in self.site.settings["permissions"]:
+                self.site.settings["permissions"].append(permission)
+
+        self.site.saveSettings()
+        self.site.updateWebsocket(permission_added=permission)
+
+        self.response(to, "ok")
+
+        for address in addresses:
+            site = self.server.sites.get(address)
+            if not site:
+                # Spawn gevent thread
+                thread = gevent.spawn(self.server.site_manager.need, address)
```

Comment:
Apparently running `need` in the same thread caused some blocking, though I can remember neither the reproduction steps nor why `join` is used here.

Reply:
We don't need to wait for the result if we don't want to process it.

If only one site is added it does not make any difference, but in case of multiple sites it will be slower with join, because it will wait for the result (complete site download) before starts downloading next site.

Reply:
Okay, I believe I would have left a better comment if I knew `thread.join` is important, so I'm going to remove it.